packages/components/src/dimension-control/index.tsx

```diff
@@ -0,0 +1,103 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Icon from '../icon';
+import SelectControl from '../select-control';
+import sizesTable, { findSizeBySlug } from './sizes';
+import type { DimensionControlProps, Size } from './types';
+import type { SelectControlProps } from '../select-control/types';
+
+/**
+ * `DimensionControl` is a component designed to provide a UI to control spacing and/or dimensions.
+ *
+ * This feature is still experimental. “Experimental” means this is an early implementation subject to drastic and breaking changes.
+ *
+ * ```jsx
+ * import { useState } from 'react';
+ * import { __experimentalDimensionControl as DimensionControl } from '@wordpress/components';
+ *
+ * export default function MyCustomDimensionControl() {
+ * 	const [ paddingSize, setPaddingSize ] = useState( '' );
+ *
+ * 	return (
+ * 		<DimensionControl
+ * 			label={ 'Padding' }
+ * 			icon={ 'desktop' }
+ * 			onChange={ ( value ) => setPaddingSize( value ) }
+ * 			value={ paddingSize }
+ * 		/>
+ * 	);
+ * }
+ * ```
+ */
+export function DimensionControl( props: DimensionControlProps ) {
+	const {
+		label,
+		value,
+		sizes = sizesTable,
+		icon,
+		onChange,
+		className = '',
+	} = props;
+
+	const onChangeSpacingSize: SelectControlProps[ 'onChange' ] = ( val ) => {
+		/* TODO: We know `val` is going to be a string (and not an array of strings) because
+		we don't pass `multiple` to `SelectControl`. Reevaluate if we can get rid of the type cast
+		after https://github.com/WordPress/gutenberg/pull/47390 is finished. */
+		const theSize = findSizeBySlug( sizes, val as string );
```

Comment:
Last request 😅 

With #47351 merged, we should be able to remove this type cast?

Something like

```suggestion
	const onChangeSpacingSize = ( val: string ) => {
		const theSize = findSizeBySlug( sizes, val );
```

Reply:
Good call! Should be fixed with https://github.com/WordPress/gutenberg/pull/47351/commits/b74c44a1a05ebfce099c2e2bc497eca1febe97b7 :)